lib/haml/attribute_builder.rb

```diff
@@ -12,6 +15,7 @@ def build_attributes(is_html, attr_wrapper, escape_attrs, hyphenate_data_attrs,
             data_attributes = attributes.delete(key)
             data_attributes = flatten_data_attributes(data_attributes, '', join_char)
             data_attributes = build_data_keys(data_attributes, hyphenate_data_attrs, key)
+            verify_attribute_names!(data_attributes.keys)
```

Comment:
> https://github.com/haml/haml/issues/891#issuecomment-276227989
> I can’t see any way of restricting that processing to just “dynamic” attributes though, so it may involve a performance hit.

This is the way that restricts processing to just dynamic attributes.